documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md 

```diff
@@ -0,0 +1,82 @@
+
+The catalog has a DAG documentation generator that extracts the docstrings from the DAGs and creates a page with the documentation. We can use the same approach to create a documentation page for the media properties. The DAG doc generator runs the code inside docker to extract DAG information. In this project, we can simplify and use the `ast` module to parse static code and extract the docstrings to make the checks faster. This will, however, mean that some of the settings will not be picked up and will have to be updated during the generation. One example of this is the `nullable` field in the `Column` class: it falls back to the value of `not nullable` during the object initiation.
+
+## Outlined Steps
+
+### `just` script
+
+The script parses the SQL DDL files, Python files and markdown files to generate a markdown page with a table and long-form notes. This script can `diff` with the existing documentation page (created in step 4 below) for the `precommit` check and post a note about how to update the docs if differences are detected.
+
+### 1. SQL DDL parsing
+We will use the names of the properties as they are in the upstream database, in the same order as they are in the `image_view`/`audio_view` materialized views. These views are selected over the main `image` and `audio` tables because they are used by the API during data refresh and contain a complete list of properties.
+
+A Python script is written to extract the information from the SQL files into a mapping with the name of the media property and the following values: SQL datatype, and SQL constraints. More data will be added to this mapping in the further steps.
+
+### 2. Parsing of the Python code
+The `Column` class and its child classes describe the validations we use to write the data collected by provider scripts. The `add_item` method has docstrings with short descriptions of what each property is. The information from these items will be added into the mapping from step 1.
+
+### 3. Adding the long-form information from the markdown file
+The markdown file will be parsed and split into the preamble, postamble (???) and the information about each data property. This information would be added to the `description` in the property mapping.
+
+### 4. Writing the markdown file
+The mapping will be used to create a new markdown file: the Preamble, the tables for Audio and Image, the long-form document from step 3.
+
+### Information to include in `media_properties.md` for each property
+Media types: Image, Audio
+Short description
+Names used in the provider scripts, if different from the database name
+Shape of the object (if it is not a simple type)
+Selection criteria
+Normalization and validation performed in the `MediaStore` class and the relevant `Column`
+Inconsistencies in the database data
+See sample for `identifier` and `tags` in `sample_media_properties.md`
```

Comment:
You mentioned earlier that some of this data would need to be updated manually. Do we know which parts?

I have a couple of questions about the manual input as well:
* How will we ensure that the `just` script doesn't overwrite that data when it auto-generates the documentation?
* How will we ensure that this gets updated as fields change or are added? Will this just be a checklist item on the PR list, for example, or did you have some automation in mind?

Reply:
This project will add `media_doc_gen` folder to `/catalog/utilities`. This will have the following files:
- `general.md`. Manually written pre- and post- ambles.
- `media_properties.md`. Manually written markdown file that will describe each media property. See examples in the [comment below](https://github.com/WordPress/openverse/pull/1904#discussion_r1179913440)
- `media_doc_generation.py`

It will also add `generate-media-docs` just script that would run `media_doc_generation.py` and create a `media_properties_doc.md` (the name needs to be improved) from the data that is collected in steps 1-3. This will be compared to `/documentation/catalog/media_properties_doc.md` that contains the most up-to-date information, and fail CI if there are differences.

Reply:
Oh I see, this comment really helped me understand the structure here. So there is one, final `media_properties_doc.md` file that’s generated by the `media_doc_generation.py` script, and its structure is:
* manually written preamble (pulled from `general.md`)
* media properties (the script will pull information for each property from docstrings and other files, and then ALSO parse out the manually written information from `media_properties.md` and add them in)
* manually written postamble (pulled from `general.md`)

So the final output is a single markdown file that includes both the manually written info, and the parsed information. Very cool!

Could you add this description of the files to the doc, maybe right before the Outlined Steps? This is really useful!